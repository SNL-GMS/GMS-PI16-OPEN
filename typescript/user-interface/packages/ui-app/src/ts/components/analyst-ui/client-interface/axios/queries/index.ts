import * as EventQuery from './event-query';
import * as QcMaskQuery from './qc-mask-query';
import * as SignalDetectionQuery from './signal-detection-query';
import * as WaveformQuery from './waveform-query';
import * as WorkflowIntervalQuery from './workflow-interval-query';
import * as WorkflowQuery from './workflow-query';
import * as WorkflowQueryUtil from './workflow-query-util';

export {
  EventQuery,
  QcMaskQuery,
  SignalDetectionQuery,
  WaveformQuery,
  WorkflowIntervalQuery,
  WorkflowQuery,
  WorkflowQueryUtil
};
