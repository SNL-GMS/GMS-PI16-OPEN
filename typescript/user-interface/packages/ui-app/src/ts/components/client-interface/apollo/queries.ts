import { ConfigurationQueries } from '@gms/common-graphql';
import { graphql } from 'react-apollo';

// ----- UI Configuration Queries ------

/**
 * Returns a wrapped component providing the `UIConfiguration` query.
 *
 * @export
 * @template T defines the component base props required
 * @returns the wrapped component
 */
// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export function graphqlUIConfigurationQuery() {
  return graphql(ConfigurationQueries.uiConfigurationQuery, { name: 'uiConfigurationQuery' });
}
