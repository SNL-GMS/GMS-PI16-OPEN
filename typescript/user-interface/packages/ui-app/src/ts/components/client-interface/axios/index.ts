import * as Mutations from './mutations';
import * as Queries from './queries';
import * as Subscriptions from './subscriptions';

export { Queries };
export { Mutations };
export { Subscriptions };
