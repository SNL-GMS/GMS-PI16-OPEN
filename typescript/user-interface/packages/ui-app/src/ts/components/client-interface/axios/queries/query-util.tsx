/* eslint-disable react/prop-types */
import { Timer } from '@gms/common-util';
import Axios, { AxiosError, AxiosRequestConfig, AxiosResponse } from 'axios';
import cloneDeep from 'lodash/cloneDeep';
import * as React from 'react';
import { QueryControls } from 'react-apollo';
import {
  QueryCache,
  QueryFunction,
  QueryKey,
  QueryResult,
  useQuery,
  UseQueryObjectConfig
} from 'react-query';

import { defaultRequestTransformers, defaultResponseTransformers } from '../axios-transformers';

/**
 * Defines type for the handler for axios queries.
 */
export type AxiosQueryHandler = (
  response: AxiosResponse,
  key?: string,
  requestConfig?: AxiosRequestConfig
) => unknown;

/**
 * Default query handler for React Query using Axios underneath.
 */
const defaultHandleQuery: AxiosQueryHandler = (response: AxiosResponse) => response.data;

/**
 * Default query implementation for React Query using Axios underneath.
 *
 * @param handleQuery (optional) query handler - allows for intercepting the data before returning and caching
 */
export function defaultQuery(handleQuery: AxiosQueryHandler = defaultHandleQuery) {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  return async (key: string, requestConfig: AxiosRequestConfig): Promise<any> => {
    try {
      Timer.start(`[axios]: query ${key} ${JSON.stringify(requestConfig.data)}`);
      const response = await Axios.request({
        ...requestConfig,
        // apply the default response transformers; unless the request config specifies its own
        transformResponse: requestConfig.transformResponse ?? defaultResponseTransformers,
        // apply the default request transformers; unless the request config specifies its own
        transformRequest: requestConfig.transformRequest ?? defaultRequestTransformers
      });
      Timer.end(`[axios]: query ${key} ${JSON.stringify(requestConfig.data)}`);
      return handleQuery(response, key, requestConfig);
    } catch (e) {
      // eslint-disable-next-line no-console
      console.error(
        `Failed Axios request: ${JSON.stringify(requestConfig)} : ${JSON.stringify(e)}`
      );
      throw e;
    }
  };
}

/**
 * The ReactQuery Cache used by the whole project
 */
export const queryCache = new QueryCache({
  defaultConfig: {
    queries: {
      queryFn: defaultQuery()
    }
  }
});

/**
 * Wraps the child component and injects (into a prop with the given propName)
 * the results of the query
 * represented by the provided queryConfig.
 *
 * @param propName the string name of the prop that should be injected
 * @param queryConfig the configuration for this query
 */
export function withQuery<TResult, TError>(
  propName: string,
  queryConfig: UseQueryObjectConfig<TResult, TError>
): React.FunctionComponent {
  return function InjectedWithQuery(props) {
    const result = useQuery<TResult, TError>(queryConfig);

    /**
     * We store the child in a variable so we can check its type with the
     * isValidElement type guard below.
     */
    const { children } = props;
    const additionalProps = {
      [propName]: result
    };
    /**
     * Verify that the children are indeed a React Element (not just a node)
     */
    if (React.isValidElement(children)) {
      /**
       * React.cloneElement is used to inject the new props into the child. This injects the
       * acknowledgeStationsByName function as a prop into the child.
       * CloneElement should be reasonably performant.
       * See https://stackoverflow.com/questions/54922160/react-cloneelement-in-list-performance
       */
      return <>{React.cloneElement(children, additionalProps)}</>;
    }
    // eslint-disable-next-line no-console
    console.error('WithQuery requires a valid React node as a child');
    return null;
  };
}

/**
 * A higher-higher order component that builds a builder for a query. Used
 * to build a function to use with compose to inject query results into the provided
 * react component.
 * Usage:
 * exampleQueryInjector = buildQueryInjector({
 *   queryKey: 'example',
 *   queryFn: () => {
 *     const { data } = await Axios.get(MOCK_ENDPOINT_ROOT + '/userProfile');
 *     return data;
 *   }
 * })
 * compose(something, exampleQueryInjector);
 *
 * @param queryConfig a definition of the query, including
 * queryKey: a serializable object that serves as the key (order matters in the case of an array)
 * queryFn: the function that actually makes the query call
 * config?: any ReactQuery configuration needed
 */
export function buildQueryInjector<TWrapped, TResult, TError>(
  propName: string,
  queryConfig: UseQueryObjectConfig<TResult, TError>
): (
  Component: React.ComponentClass | React.FunctionComponent
) => React.FunctionComponent<TWrapped & { [x: string]: QueryResult<TResult, TError> }> {
  const QueryInjector = withQuery<TResult, TError>(propName, queryConfig);
  // eslint-disable-next-line react/display-name
  return Component => props => (
    <QueryInjector>
      {/* eslint-disable-next-line react/jsx-props-no-spreading */}
      <Component {...props} />
    </QueryInjector>
  );
}

/**
 * Checks if a query is in a loading state.
 *
 * @param query either a react-apollo or react-query query
 * @returns true if the query is loading
 */
export const isQueryLoading = (
  query: QueryControls<unknown, Record<string, AxiosError>> & QueryResult<unknown, AxiosError>
): boolean => query.loading || query.isLoading;

type StructuredQueryKey = [string, AxiosRequestConfig];

function isStructuredQueryKey(qKey: QueryKey): qKey is StructuredQueryKey {
  if (
    Array.isArray(qKey) &&
    qKey.length === 2 &&
    typeof qKey[0] === 'string' &&
    qKey[1] &&
    Object.prototype.hasOwnProperty.call(qKey[1], 'url')
  ) {
    return true;
  }
  return false;
}

/**
 * Creates a queryConfig with the provided data that may be provided to ReactQuery's
 * fetch/query functions.
 *
 * @throws if the baseConfig does not have a queryKey of the format [string, AxiosRequestConfig]
 * @param baseConfig the default configuration onto which we will add the data. Expects
 * to have a queryKey of the structure: [string, AxiosRequestConfig]
 * @param data the data to add to the AxiosRequestConfig in the baseConfig.
 */
export function createQueryConfig<TData>(
  baseConfig: UseQueryObjectConfig<TData, AxiosError>,
  data: unknown,
  queryFn?: QueryFunction<TData>
): UseQueryObjectConfig<TData, AxiosError> {
  const newConfig = cloneDeep(baseConfig);

  const qKey = newConfig.queryKey;
  if (isStructuredQueryKey(qKey)) {
    const requestWithData = cloneDeep(qKey[1]);
    requestWithData.data = data;
    newConfig.queryKey = [qKey[0], requestWithData];
    if (queryFn) {
      newConfig.queryFn = queryFn;
    }
    return newConfig;
  }
  throw Error(
    'createQueryConfig expects to have a baseConfig with a queryKey of type [string, AxiosRequestConfig]'
  );
}
