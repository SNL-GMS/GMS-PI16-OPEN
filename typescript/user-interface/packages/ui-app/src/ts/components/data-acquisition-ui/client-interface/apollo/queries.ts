import { ProcessingStationQueries } from '@gms/common-graphql';
import { graphql } from 'react-apollo';

// ----- Station Processing Queries ------

/**
 * Returns a wrapped component providing the `defaultStationsQuery` query.
 *
 * @export
 * @returns the wrapped component
 */
export function graphqlDefaultProcessingStationsQuery() {
  return graphql(ProcessingStationQueries.defaultProcessingStationsQuery, {
    name: 'defaultStationsQuery'
  });
}
