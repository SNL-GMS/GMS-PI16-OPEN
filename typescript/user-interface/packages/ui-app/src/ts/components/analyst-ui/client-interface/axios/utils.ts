import { CommonTypes, WorkflowTypes } from '@gms/common-model';
import { MILLISECONDS_IN_SECOND, toEpochSeconds } from '@gms/common-util';
import Immutable from 'immutable';
import moment from 'moment';

import { queryCache } from '~components/client-interface';

import { getWorkflowIntervalQueryConfig } from './queries/workflow-interval-query';

// TODO once our query key logic is cleaned up, will not need stageNames and timeRange
export const cleanUpWorkflow = (
  latestIntervalEndTime: number,
  operationalTimeDuration: number,
  stageNames: string[],
  timeRange: CommonTypes.TimeRange
): void => {
  const startTime = toEpochSeconds(
    moment((latestIntervalEndTime - operationalTimeDuration) * MILLISECONDS_IN_SECOND)
      .utc()
      .startOf('hour')
      .toISOString()
  );
  const { queryKey } = getWorkflowIntervalQueryConfig(stageNames, timeRange);
  let cachedWorkflowIntervalResultMap: Immutable.Map<
    string,
    WorkflowTypes.StageInterval[]
  > = queryCache.getQueryData(queryKey);
  // TODO maybe optimize with bailing once finding a valid endTime
  let foundStaleEntries = false;
  cachedWorkflowIntervalResultMap.forEach((entries, key) => {
    const newEntries = entries
      .map(entry => {
        if (entry.startTime < startTime) {
          foundStaleEntries = true;
          return undefined;
        }
        return entry;
      })
      .filter(entry => entry !== undefined);
    cachedWorkflowIntervalResultMap = cachedWorkflowIntervalResultMap.set(key, newEntries);
  });
  if (foundStaleEntries) {
    queryCache.setQueryData(queryKey, cachedWorkflowIntervalResultMap);
  }
};
