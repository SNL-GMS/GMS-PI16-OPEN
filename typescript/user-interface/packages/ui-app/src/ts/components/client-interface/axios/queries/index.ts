import * as ProcessingAnalystConfigurationQuery from './processing-analyst-configuration-query';
import * as ProcessingCommonConfigurationQuery from './processing-common-configuration-query';
import * as OperationalTimePeriodConfigurationQuery from './processing-operational-time-period-configuration-query';
import * as ProcessingStationGroupNamesConfigurationQuery from './processing-station-group-names-configuration-query';
import * as QueryUtils from './query-util';
import * as StationGroupQuery from './station-group-query';
import * as StationGroupsDefinitionQuery from './station-groups-definition-query';
import * as StationQueryUtil from './station-query-util';
import * as StationsChangeTimesQuery from './stations-change-times-query';
import * as StationsDefinitionQuery from './stations-definition-query';
import * as SystemMessageDefinitionQuery from './system-message-definition-query';
import * as UserProfileQuery from './user-profile-query';

export {
  OperationalTimePeriodConfigurationQuery,
  ProcessingAnalystConfigurationQuery,
  ProcessingCommonConfigurationQuery,
  ProcessingStationGroupNamesConfigurationQuery,
  QueryUtils,
  StationGroupQuery,
  StationGroupsDefinitionQuery,
  StationQueryUtil,
  StationsChangeTimesQuery,
  StationsDefinitionQuery,
  SystemMessageDefinitionQuery,
  UserProfileQuery
};
