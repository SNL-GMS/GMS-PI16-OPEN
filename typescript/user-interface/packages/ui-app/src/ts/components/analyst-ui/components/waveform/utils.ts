import { CommonTypes, EventTypes } from '@gms/common-model';
import { WeavessTypes } from '@gms/weavess-core';

import { isSdInstantMeasurementValue } from '~analyst-ui/common/utils/instance-of-util';

export interface Offset {
  channelId: string;
  offset: number;
}

/**
 * Calculate offsets based on station with earliest arrival.
 */
export const calculateOffsets = (
  featurePredictions: EventTypes.FeaturePrediction[],
  baseChannelId: string,
  phaseToOffset: string
): Offset[] => {
  if (!featurePredictions || featurePredictions.length <= 0) {
    return [];
  }
  const filteredPredictions: EventTypes.FeaturePrediction[] = featurePredictions.filter(
    fp => fp.phase === phaseToOffset
  );

  // If no FPs were found for the phase selected, return nothing
  if (filteredPredictions.length <= 0) {
    return [];
  }

  filteredPredictions.sort((a, b) => {
    if (
      isSdInstantMeasurementValue(a.predictedValue) &&
      isSdInstantMeasurementValue(b.predictedValue)
    ) {
      const aValue = a.predictedValue.value;
      const bValue = b.predictedValue.value;
      return aValue - bValue;
    }
    return 0;
  });

  const offsets: Offset[] = [];
  const baseChannelFeaturePrediction = filteredPredictions.find(
    fp => fp.channelName === baseChannelId
  );

  if (
    baseChannelFeaturePrediction &&
    isSdInstantMeasurementValue(baseChannelFeaturePrediction.predictedValue)
  ) {
    const baseTime: number = baseChannelFeaturePrediction.predictedValue.value;

    filteredPredictions.forEach(sortedFeaturePrediction => {
      if (isSdInstantMeasurementValue(sortedFeaturePrediction.predictedValue)) {
        const offset: Offset = {
          channelId: sortedFeaturePrediction.channelName,
          offset: baseTime - sortedFeaturePrediction.predictedValue.value
        };
        offsets.push(offset);
      }
    });
  }
  return offsets;
};

/**
 * TODO: Remove if/when we convert the UI TimeRange to use the same property keys.
 * Converts a UI time range to the Weavess format.
 *
 * @param timeRange a time range in the common model format
 * @returns a timeRange in the weavess format
 */
export const convertToWeavessTimeRange = (
  timeRange: CommonTypes.TimeRange
): WeavessTypes.TimeRange => ({
  startTimeSecs: timeRange.startTimeSecs,
  endTimeSecs: timeRange.endTimeSecs
});
