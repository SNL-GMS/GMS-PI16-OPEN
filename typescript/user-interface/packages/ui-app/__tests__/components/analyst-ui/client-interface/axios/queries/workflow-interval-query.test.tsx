import { CommonTypes, WorkflowTypes } from '@gms/common-model';
import { AxiosError } from 'axios';
import Immutable from 'immutable';
import { UseQueryObjectConfig } from 'react-query';

import {
  fetchWorkflowIntervals,
  getWorkflowIntervalQueryConfig,
  useWorkflowIntervalQuery,
  workflowIntervalQueryConfig
} from '../../../../../../src/ts/components/analyst-ui/client-interface/axios/queries/workflow-interval-query';
import { queryCache } from '../../../../../../src/ts/components/client-interface/axios/queries/query-util';
import { expectQueryHookToMakeAxiosRequest } from '../../../../../utils/query-mutation-test-utils';

describe('Workflow Query', () => {
  const timeRange: CommonTypes.TimeRange = {
    startTimeSecs: 0,
    endTimeSecs: 100
  };
  describe('starts with', () => {
    it('a hook for accessing the workflow interval query defined', () => {
      expect(useWorkflowIntervalQuery).toBeDefined();
    });
    it('a query config defined', () => {
      expect(workflowIntervalQueryConfig).toBeDefined();
    });
    it('fetchWorkflow intervals defined', () => {
      expect(fetchWorkflowIntervals).toBeDefined();
    });

    it('getWorkflowIntervalQueryConfig defined', () => {
      expect(fetchWorkflowIntervals).toBeDefined();
    });
  });

  it('getWorkflowIntervalQueryConfig returns valid config', () => {
    const config: UseQueryObjectConfig<
      Immutable.Map<string, WorkflowTypes.StageInterval[]>,
      AxiosError
    > = getWorkflowIntervalQueryConfig([], timeRange);
    expect(config).toMatchSnapshot();
  });

  it('fetchWorkflow returns the expected result', async () => {
    const queryCacheMock = {
      fetchQuery: jest.fn()
    };
    Object.assign(queryCache, queryCacheMock);

    queryCacheMock.fetchQuery.mockReturnValueOnce('Query can be fetched');
    const result = await fetchWorkflowIntervals([], timeRange);

    expect(result).toEqual('Query can be fetched');
  });

  it('hook queries for workflow interval query', async () => {
    const useTestHook = () => useWorkflowIntervalQuery(['stageName1', 'stageName2'], timeRange);
    await expectQueryHookToMakeAxiosRequest(useTestHook);
    expect(true).toBeTruthy();
  });

  it('fails to fetch workflow, error thrown', async () => {
    expect.assertions(1);
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const err: any = new Error('Nope');
    const errWrapped: Error = new Error(err);
    const failQueryCacheMock = {
      fetchQuery: jest.fn(() => {
        throw err;
      })
    };
    Object.assign(queryCache, failQueryCacheMock);

    await fetchWorkflowIntervals([], timeRange).catch(e => {
      expect(e).toEqual(errWrapped);
    });
  });
});
