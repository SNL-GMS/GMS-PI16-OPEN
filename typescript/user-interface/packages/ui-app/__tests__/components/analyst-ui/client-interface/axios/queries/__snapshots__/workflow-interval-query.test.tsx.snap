// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Workflow Query getWorkflowIntervalQueryConfig returns valid config 1`] = `
Object {
  "config": Object {
    "enabled": false,
    "refetchOnWindowFocus": false,
    "staleTime": Infinity,
  },
  "queryFn": [Function],
  "queryKey": Array [
    "getWorkflowIntervals",
    Object {
      "data": Object {
        "endTime": 100,
        "stageIds": Array [],
        "startTime": 0,
      },
      "headers": Object {
        "accept": "application/json",
        "content-type": "text/plain",
      },
      "method": "post",
      "proxy": false,
      "responseType": "json",
      "timeout": 180000,
      "url": "http://localhost/workflow-manager-service/workflow-manager/interval/stage/query/ids-timerange",
    },
  ],
}
`;
